### PR TITLE
Improve RigidDynamicBody contacts in 2D and 3D

### LINKED ISSUE #2092: Pile of RigidBody2Ds results in overlapping and shaking
Dump a bunch of rigibody2d's with rectangle collision shapes into a pit so they pile on top of each other, and the ones at the bottom are forced into each other.  They overlap significantly, and constantly shake around as they try to resolve their colliisions.

Example:
https://copy.com/YWldXOFC18N1oYRk

### LINKED ISSUE #29303: Rigid bodies using inertia become unstable.
**Godot version:**

8c923fc

**OS/device including version:**

Win10 64-bit

**Issue description:**

Decided to revisit these to see if they're usable, and when trying to change the inertia in script, I got rather bad results.  This video shows some pairs of shapes, with and without custom inertia.  

The one of the left is without, and is the expected result, the copy on the right is with some inertia set.

ezgif com-video-to-gif (1)

**Steps to reproduce:**
Make some RigidBody2Ds, and in script set their inertia property.  Values closer to 1 seem to be the most unstable.

**Minimal reproduction project:**

[RigidBody Bug.zip](https://github.com/godotengine/godot/files/3235733/RigidBody.Bug.zip)

### LINKED ISSUE #34215: 2D Physics Collision is 'Squishy'
**Godot version:**
3.1.2.stable.official

**OS/device including version:**
Windows 10

**Issue description:**
When two a `RigidBody2D` collides with a `StaticBody2D`, they 'squish' into each other and then the `RigidBody2D` slowly returns to the edge of the collision over a few frames. This makes the collision not feel quite so 'rigid' but instead like the `StaticBody2D` is squishy and allows other bodies to enter it temporarily. I have not checked to see if this behavior presents in 3D Nodes as well.

collision

**Steps to reproduce:**
Create a 2D scene with a `RigidBody2D` and a `StaticBody2D`. Have the `RigidBody2D` collide into the other body with some velocity. You'll observe the `RigidBody2D` enter the `StaticBody2D`.

**Minimal reproduction project:**
Zipped Project, open at run the scene.
[Bouncy.zip](https://github.com/godotengine/godot/files/3937910/Bouncy.zip)

### PR DESCRIPTION
**Pull request by pouleyKetchoupp**
Improved rigid body contacts and physics server contact settings, separated in two commits to make it easier to read.

Causes slight compatibility breakage because of changing physics server parameters and their default values.

**First commit: main changes in rigid body contacts**
Changed the algorithm for solving contacts to keep previous contacts as long as they are under the max separation threshold to keep contact impulses more consistent and contacts more stable.

Also made 2D consistent with 3D and changed some default parameters:
-Contact bias is now 0.8 instead of 0.3 to avoid springy contacts
-Solver iterations are 16 instead of 8 by default for better stability

Performance considerations:
Tested with stress tests that include lots of contacts from overlapping bodies.
3D: There's no measurable difference in performance.
2D: Performance is a bit lower (close to 10% slower in extreme cases)
The benefit for 2D physics to be much more stable outweighs the slight decrease in performance, and this could be alleviated by changing the algorithm to use jacobians for contact solving to help with cache efficiency and memory allocations.

**Second commit: Update space parameters in 2D and 3D**
Clarified space parameters for contacts and added missing ones.

List of changes:
-Add contact bias to space parameters
-Add solver iterations to space parameters, instead of a specific physics server function
-Renamed `BODY_MAX_ALLOWED_PENETRATION` to `CONTACT_MAX_ALLOWED_PENETRATION` to make it consistent with other contact parameters

---

Tests with stack of boxes from the 2D and 3D physics test demos:
https://github.com/godotengine/godot-demo-projects/pull/638

Now they work fine in 2D and 3D, even with the increased 2D gravity in 4.0.

| | Before | After |
| ------- | ------- | ------- |
| Pile of boxes 2D |  |  |
| Pile of boxes 3D |  |  |
| Push boxes 2D |  |  |

---

Fixed bugs:

Fixes #2092 - Pile of RigidBody2Ds results in overlapping and shaking
C:\Users\bourr\Downloads\godot-projects\test_rigidbody-4.0
4.0 project: [test_rigidbody-4.0.zip](https://github.com/godotengine/godot/files/7651842/test_rigidbody-4.0.zip)
Same project as original MVP, with gravity set to 98 to match the default from 3.x

Fixes #34215 - 2D Physics Collision is 'Squishy'
4.0 project: [Bouncy-4.0.zip](https://github.com/godotengine/godot/files/7651846/Bouncy-4.0.zip)

Fixes #29303 - Rigid bodies using inertia become unstable.
4.0 project: [RigidBody.Bug-4.0.zip](https://github.com/godotengine/godot/files/7651849/RigidBody.Bug-4.0.zip)
Same project as MVP, with more reasonable values for inertia to help with stability (no more than 20 times the default amount of rotation, which was still causing issues in 3.x).
Works well now, even with gravity 10 times higher in 4.0.

### CODE REVIEW

File: servers/physics_2d/godot_body_pair_2d.h [hunk 1]
```diff
@@ -62,13 +62,14 @@ class GodotBodyPair2D : public GodotConstraint2D {
 		real_t acc_normal_impulse = 0.0; // accumulated normal impulse (Pn)
 		real_t acc_tangent_impulse = 0.0; // accumulated tangent impulse (Pt)
 		real_t acc_bias_impulse = 0.0; // accumulated normal impulse for position bias (Pnb)
+		real_t acc_bias_impulse_center_of_mass = 0.0; // accumulated normal impulse for position bias applied to com
```
Comment (fabriceci) [inline servers/physics_2d/godot_body_pair_2d.h:65]: Shouldn't inline comments also start with a capital letter and end with a point?

**Review by fabriceci — COMMENTED** [review]
I tried with a custom tests, there is a big improvement, on piles indeed but on the overall stability.

~~@pouleyKetchoupp as you can see, it's maybe not related to the PR, but the contact points are not displayed after. Can you just verify that on your side?~~ not related, fixed few days ago by #55486

Before:

https://user-images.githubusercontent.com/6397893/144765070-70e2617b-f9dc-4c02-9454-7a72cc21188c.mp4

After:

https://user-images.githubusercontent.com/6397893/144765079-bd69471b-0720-4914-9677-d1434387014d.mp4

**Review by fabriceci — APPROVED** [review]

**Review by reduz — APPROVED** [review]

File: servers/physics_2d/godot_body_pair_2d.h (hunk 1 above)
Comment (pouleyKetchoupp) [reply servers/physics_2d/godot_body_pair_2d.h:65]: Yeah these comments should follow the same rules, but for now I've just copied it the same way. Physics servers have more styling issues like that and will require a pass to fix that at some point :)

**Comment by akien-mga** [comment]
Thanks! :tada:

**Pull request merged**